hail/hail/src/is/hail/expr/ir/IR.scala

```diff
@@ -1063,12 +1055,30 @@ package defs {
     returnType: Type,
     errorID: Int,
   ) extends IR {
-    var conversion: (Seq[Type], Seq[IR], Int) => IR = _
-    var inline: Boolean = _
+    private var fetchedFromRegistry: Boolean = false
+
+    private def fetchFromRegistry(): Unit = {
+      val ((_, _, _, inline), impl) =
+        IRFunctionRegistry.lookupIR(function, typeArgs, args.map(_.typ)).get
+      _body = impl(typeArgs, refs, errorID).deepCopy()
+      _inline = inline
+      fetchedFromRegistry = true
+    }
+
+    private var _body: IR = _
+    private var _inline: Boolean = _
```

Comment:
This feels exactly what lazy vals are for
```scala
[private] lazy val fetch: (IR, Boolean) = {
  val ((_, _, _, inline), impl) =
    IRFunctionRegistry.lookupIR(function, typeArgs, args.map(_.typ)).get
  (impl(typeArgs, refs, errorID).deepCopy(), inline)
}

Reply:
I think this could even work as:

```scala
lazy val (body, inline): (IR, Boolean) = ...
```

Reply:
Yeah, the only reason I didn't use lazy vals was to avoid performing the registry lookup twice. But I think you're right I can still achieve that with a lazy val. Good call!